### PR TITLE
[WIP] Serde Module for Import/Export of models between Onnx and Mxnet

### PR DESCRIPTION
**Pull request by anirudhacharya**
## Description ##
ImportExport module based on this design doc - https://cwiki.apache.org/confluence/display/MXNET/Proposal%3A+ImportExport+module.
This PR is a work in progress, we are just sending out this skeleton code to get views on the api specification, general code structure, and any additional dependencies it may pull into the project. As this module is also being discussed currently in the dev list, it is open to change in the design and specification of the project.

### Changes ###
- [ ] Import from Onnx.
- [ ] Export to Onnx.

## Comments ##
- I have currently added this project under contrib, as it is used for experimental features which can later be moved elsewhere.

@spidyDev @Roshrini @lupesko @szha

### CODE REVIEW

File: python/mxnet/contrib/__init__.py [hunk 1]
```diff
@@ -28,3 +28,4 @@
 from . import tensorboard
 
 from . import text
+from . import serde
```
Comment (piiswrong) [inline python/mxnet/contrib/__init__.py]: serde is a strange name.
I would rather call it serialization

File: python/mxnet/contrib/serde/_export/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import onnx
+
+def export_model(sym, params):
+    pass
```
Comment (piiswrong) [inline python/mxnet/contrib/serde/_export/__init__.py]: ?

File: python/mxnet/contrib/__init__.py (hunk 1 above)
Comment (piiswrong) [inline python/mxnet/contrib/__init__.py]: https://github.com/dmlc/

File: python/mxnet/contrib/serde/_import/backend.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+from .... import ndarray as nd
+
+# Using these functions for onnx test infrastructure.
+# Implemented by following onnx docs guide:
+# https://github.com/onnx/onnx/blob/master/docs/Implementing%20an%20ONNX%20backend.md
+# MXNetBackend class will take an ONNX model with inputs, perform a computation,
+# and then return the output.
+
+class MXNetBackend(Backend):
+    """MXNet backend for ONNX"""
```
Comment (piiswrong) [inline python/mxnet/contrib/serde/_import/backend.py]: what's this?

File: python/mxnet/contrib/serde/_import/common.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+# Derived from Apache 2.0 licensed common.py file from DMLC NNVM:
+# https://github.com/dmlc/nnvm/blob/3da53e46db57c438b05fbebe8aa332ee8c5994d1/python/nnvm/frontend/common.py
+
+# coding: utf-8
+# pylint: disable=invalid-name,no-self-use,too-many-branches,too-few-public-methods,too-many-arguments
+"""Shared functions and classes for frontends."""
+from __future__ import absolute_import as _abs
+from ....base import string_types
+
+class Renamer(object):
```
Comment (piiswrong) [inline python/mxnet/contrib/serde/_import/common.py]: This class looks unnecessary

File: python/mxnet/contrib/serde/_import/tests/onnx_backend_test.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import unittest
+
+import onnx.backend.test
+from onnx_mxnet import backend as mxnet_backend
+
+# This is a pytest magic variable to load extra plugins
+pytest_plugins = 'onnx.backend.test.report'
```
Comment (piiswrong) [inline python/mxnet/contrib/serde/_import/tests/onnx_backend_test.py]: tests should be put in tests/python/unittests

File: python/mxnet/contrib/serde/_import/import_onnx.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        begin = new_attr.get('begin')
+        end = new_attr.get('end')
+        axes = new_attr.get('axis', tuple(range(len(begin))))
+        slice_op = symbol.slice_axis(inputs[0], axis=axes[0], begin=begin[0], end=end[0])
+        if len(axes) > 1:
+            for i, axis in enumerate(axes):
+                slice_op = symbol.slice_axis(slice_op, axis=axis, begin=begin[i], end=end[i])
+        return slice_op
+
+    def _fix_squeeze(self, inputs, new_attr):
```
Comment (piiswrong) [inline python/mxnet/contrib/serde/_import/import_onnx.py]: why not do this in attribute converter?

File: python/mxnet/contrib/serde/_import/import_onnx.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        out = [self._nodes[i.name] for i in graph.output]
+        if len(out) > 1:
+            out = symbol.Group(out)
+        else:
+            out = out[0]
+        return out, self._params
+
+    def run_node(self, node, device='CPU'): # pylint: disable=unused-argument
+        """Construct symbol from individual node.
+        Mainly using this function for unittests"""
```
Comment (piiswrong) [inline python/mxnet/contrib/serde/_import/import_onnx.py]: test utility should be put into test code

File: python/mxnet/contrib/serde/_export/__init__.py (hunk 2 above)
Comment (anirudhacharya) [reply python/mxnet/contrib/serde/_export/__init__.py]: Will delete this module for now.

File: python/mxnet/contrib/__init__.py (hunk 1 above)
Comment (anirudhacharya) [reply python/mxnet/contrib/__init__.py]: Will change the module name from serde to onnx.

File: python/mxnet/contrib/serde/_import/backend.py (hunk 3 above)
Comment (anirudhacharya) [reply python/mxnet/contrib/serde/_import/backend.py]: This class is used by the onnx backend test framework here - https://github.com/anirudhacharya/incubator-mxnet/blob/serde/python/mxnet/contrib/serde/_import/tests/onnx_backend_test.py

For testing the import functionality, we intend to use ONNX's backend test framework as described here - https://github.com/onnx/onnx/blob/master/docs/OnnxBackendTest.md 
This is done to ensure that tests will be shared across different frameworks and that our module will be sync with onnx's standards and definition for various operators. This class will be used by the test framework to run operators on mxnet backend.

File: python/mxnet/contrib/serde/_import/tests/onnx_backend_test.py (hunk 5 above)
Comment (anirudhacharya) [reply python/mxnet/contrib/serde/_import/tests/onnx_backend_test.py]: will change it.

File: python/mxnet/contrib/serde/_import/import_onnx.py (hunk 7 above)
Comment (anirudhacharya) [reply python/mxnet/contrib/serde/_import/import_onnx.py]: Will refactor this code along with Renamer and Attribute Converter class.

File: python/mxnet/contrib/__init__.py (hunk 1 above)
Comment (lupesko) [reply python/mxnet/contrib/__init__.py]: +1
Serde can come in later as a facade on top of onnx and coreml

File: python/mxnet/contrib/serde/_import/__init__.py [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# coding: utf-8
+"""import function"""
+import onnx
+from .import_onnx import GraphProto
+
+def import_model(model_file):
```
Comment (lupesko) [inline python/mxnet/contrib/serde/_import/__init__.py]: Propose to rename to import_onnx_model

File: python/mxnet/contrib/serde/_import/backend.py [hunk 9]
```diff
@@ -0,0 +1,131 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (lupesko) [inline python/mxnet/contrib/serde/_import/backend.py]: 2018

File: python/mxnet/contrib/serde/_import/__init__.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+# coding: utf-8
```
Comment (lupesko) [inline python/mxnet/contrib/serde/_import/__init__.py]: Please add copyright headers

File: python/mxnet/contrib/serde/_import/backend.py [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+#     http://www.apache.org/licenses/LICENSE-2.0
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+# coding: utf-8
+# pylint: disable=too-many-locals,invalid-name
+"""backend wrapper for onnx test infrastructure"""
+from collections import namedtuple
```
Comment (lupesko) [inline python/mxnet/contrib/serde/_import/backend.py]: nit: newline

File: python/mxnet/contrib/serde/_import/backend_rep.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (lupesko) [inline python/mxnet/contrib/serde/_import/backend_rep.py]: 2018, repeats in more files, please update.

File: python/setup.py [hunk 13]
```diff
@@ -28,7 +28,7 @@
 else:
     from setuptools import setup
     from setuptools.extension import Extension
-    kwargs = {'install_requires': ['numpy<=1.13.3,>=1.8.2', 'requests==2.18.4', 'graphviz==0.8.1'], 'zip_safe': False}
+    kwargs = {'install_requires': ['numpy<=1.13.3,>=1.8.2', 'requests==2.18.4', 'graphviz==0.8.1', 'onnx>=1.0.1'], 'zip_safe': False}
```
Comment (lupesko) [inline python/setup.py]: Let's make ONNX a runtime dependency, in the future we can update it to be an install time dependency.

Comment (anirudhacharya) [reply python/setup.py]: will do.

File: python/mxnet/contrib/serde/_import/backend_rep.py (hunk 12 above)
Comment (anirudhacharya) [reply python/mxnet/contrib/serde/_import/backend_rep.py]: I will update all the license and copyright statements.

File: python/mxnet/contrib/serde/_import/__init__.py (hunk 8 above)
Comment (reminisce) [reply python/mxnet/contrib/serde/_import/__init__.py]: Is `import_model` a user level API? If so, why is it registered under a hidden module `_import`?

File: python/mxnet/contrib/serde/_export/__init__.py (hunk 2 above)
Comment (reminisce) [reply python/mxnet/contrib/serde/_export/__init__.py]: Is this the determined API for exporting MXNet models to onnx format? Have you considered whether we can borrow the interface of this function in Pytorch?
https://github.com/pytorch/pytorch/blob/master/torch/onnx/__init__.py#L43

**Comment by anirudhacharya** [comment]
will raise another PR. closing this PR

**Pull request closed without merging**